oak/server/rust/oak_runtime/src/runtime.rs

```diff
@@ -88,6 +97,198 @@ impl Runtime {
+                s == ChannelReadStatus::InvalidChannel || s == ChannelReadStatus::Orphaned
+            });
+            let any_ready = statuses.iter().any(|&s| s == ChannelReadStatus::ReadReady);
+
+            if all_unreadable || any_ready {
+                return Ok(statuses);
+            }
+
+            debug!(
+                "wait_on_channels: channels not ready, parking thread {:?}",
+                oak_platform::current_thread().id()
+            );
+
+            oak_platform::park_thread();
+
+            debug!(
+                "wait_on_channels: thread {:?} re-woken",
+                oak_platform::current_thread().id()
+            );
+        }
+        Err(OakStatus::ErrTerminated)
+    }
+
+    /// Write a message to a channel. Fails with [`OakStatus::ErrChannelClosed`] if the underlying
+    /// channel has been orphaned.
+    pub fn channel_write(&self, channel: &ChannelWriter, msg: Message) -> Result<(), OakStatus> {
+        if channel.is_orphan() {
+            return Err(OakStatus::ErrChannelClosed);
+        }
+
+        {
+            let mut messages = channel.messages.write().unwrap();
+
+            messages.push_back(msg);
+        }
+
+        let mut waiting_threads = channel.waiting_threads.lock().unwrap();
+
+        // Unpark (wake up) all waiting threads that still have live references. The first thread
+        // woken can immediately read the message, and others might find `messages` is empty before
+        // they are even woken. This should not be an issue (being woken does not guarantee a
+        // message is available), but it could potentially result in some particular thread always
+        // getting first chance to read the message.
+        //
+        // If a thread is woken and finds no message it will take the `waiting_threads` lock and
+        // add itself again. Note that since that lock is currently held, the woken thread will add
+        // itself to waiting_threads *after* we call clear below as we release the lock implicilty
+        // on leaving this function.
+        for thread in waiting_threads.values() {
+            if let Some(thread) = thread.upgrade() {
+                thread.unpark();
+            }
+        }
+        waiting_threads.clear();
+
+        Ok(())
+    }
+
+    /// Thread safe. Read a message from a channel. Fails with [`OakStatus::ErrChannelClosed`] if
+    /// the underlying channel is empty and has been orphaned.
+    pub fn channel_read(&self, channel: &ChannelReader) -> Result<Option<Message>, OakStatus> {
+        let mut messages = channel.messages.write().unwrap();
+        match messages.pop_front() {
+            Some(m) => Ok(Some(m)),
+            None => {
+                if channel.is_orphan() {
+                    Err(OakStatus::ErrChannelClosed)
+                } else {
+                    Ok(None)
+                }
+            }
+        }
+    }
+
+    /// Thread safe. This function will return:
```

Comment:
(nit) we seem to use present tense elsewhere (though not super consistently). Perhaps we can do another pass of fixing this in the future though.

```suggestion
    /// Thread safe. This function returns:
```